### PR TITLE
[client/catapult] fix: add Conan 2 support

### PR DESCRIPTION
**Pull request by Wayonb**
problem: Conan 1.x is getting deprecated
solution: upgrade catapult build to use Conan 2

1. change the cmake requirement from 3.14 to 3.23.
     cmake preset requires 3.23
2. only build static lib for RocksDB on Windows
3. update Conan build-docs
4. remove the install limit for Conan
5. update the catapult dependency updater

### CODE REVIEW

File: client/catapult/docs/BUILD-conan.md [hunk 1]
```diff
@@ -13,32 +13,27 @@ Following instructions should work on Mac, Linux (Ubuntu 20.04) and Windows.
      sudo apt install build-essential git cmake ninja-build pkg-config
      ```
 
-  2. Install [Conan](https://conan.io/downloads.html).
+  2. Install latest version of [Conan](https://conan.io/downloads.html).
 
   3. Set the right C++ ABI for Conan:
```
Comment (Jaguar0625) [inline client/catapult/docs/BUILD-conan.md:18]: C++ABI => config ?

File: client/catapult/docs/BUILD-conan.md [hunk 2]
```diff
@@ -77,9 +67,19 @@ conan remote add nemtech https://conan.symbol.dev/artifactory/api/conan/catapult
 
 git clone https://github.com/symbol/symbol.git
 cd symbol/client/catapult
+```
+
+### Linux and Macos
+```sh
+conan install . --build=missing -s build_type=Release
+cd build/Release
+
```
Comment (Jaguar0625) [inline client/catapult/docs/BUILD-conan.md:76]: remove blank line

File: client/catapult/docs/BUILD-conan.md [hunk 3]
```diff
@@ -92,13 +92,13 @@ CONAN_REVISIONS_ENABLED=1 conan install .. --build missing
...
 
   ```sh
-  cmake -G "Visual Studio 17 2022" -A x64 -DUSE_CONAN=ON -DPYTHON_EXECUTABLE:FILEPATH=X:/python3x/python.exe ..
+  cmake --preset conan-default -G "Visual Studio 17 2022" -A x64 -DUSE_CONAN=ON -DPYTHON_EXECUTABLE:FILEPATH=X:/python3x/python.exe ..
   ```
 
 * Generate project files for Visual Studio 2019:
 
   ```sh
-  cmake -G "Visual Studio 16 2019"  -A x64 -DUSE_CONAN=ON -DPYTHON_EXECUTABLE:FILEPATH=X:/python3x/python.exe ..
```
Comment (Jaguar0625) [inline client/catapult/docs/BUILD-conan.md:101]: this seems like old VS version now 😅

File: client/catapult/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+		# release dependencies - openssl
+		self.options["openssl*"].shared = True
+		self.options["openssl*"].no_zlib = True
+		self.options["openssl*"].no_legacy = False
+		self.options["openssl*"].no_module = True
+
+		# release dependencies - other
+		self.options["mongo-cxx-driver*"].shared = True
+		self.options["rocksdb*"].shared = False if self.settings.os == "Windows" else True
```
Comment (Jaguar0625) [inline client/catapult/conanfile.py:70]: ` "Windows" != self.settings.os `, no need for ternary

File: client/catapult/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+				copy(self, "*.dll", dep.cpp_info.bindirs[0], dependency_path)
+				copy(self, "*.dll", dep.cpp_info.libdirs[0], dependency_path)
+			elif "Linux" == self.settings.os:
+				# on linux,  copy shared libraries into deps directory
+				copy(self, "*.so*", dep.cpp_info.libdirs[0], dependency_path)
+			elif "Macos" == self.settings.os:
+				# on macos, copy shared libraries into deps directory
+				copy(self, "*.dylib", dep.cpp_info.libdirs[0], dependency_path)
+
+		tc = CMakeToolchain(self)
```
Comment (Jaguar0625) [inline client/catapult/conanfile.py:97]: tc => toolchain

File: client/catapult/CMakeLists.txt [hunk 6]
```diff
@@ -114,3 +114,8 @@ if(EXISTS "${CMAKE_SOURCE_DIR}/internal/CMakeLists.txt")
 endif()
 
 add_subdirectory(tools)
+
+if(USE_CONAN)
+	message("--- copy conan deps ---")
+	install(DIRECTORY ${CMAKE_BINARY_DIR}/deps/ DESTINATION deps)
+endif()
```
Comment (Jaguar0625) [inline client/catapult/CMakeLists.txt:121]: why is this needed? isn't this logic in CatapultConan.generate ?

File: jenkins/catapult/catapultDependencyUpdater.py [hunk 7]
```diff
@@ -46,23 +46,23 @@ def __init__(self, versions_file, source_path):
...
-		return dependency_name in ['benchmark', 'cppzmq', 'libzmq', 'mongo-c-driver', 'mongo-cxx-driver', 'rocksdb']
+		return dependency_name in ['benchmark', 'cppzmq', 'zeromq', 'mongo-c-driver', 'mongo-cxx-driver', 'rocksdb']
 
 	def _get_conan_latest_version(self, name):
 		query = f'{name}/*' if not self._is_remote_nemtech(name) else f'{name}/*@{NEMTECH_REMOTE_NAME}/stable'
 		remote_name = NEMTECH_REMOTE_NAME if self._is_remote_nemtech(name) else 'conancenter'
-		output, _ = dispatch_subprocess(['conan', 'search', query, '--raw', '--remote', remote_name])
-		version_list = [line.split('/')[1].split('@')[0] for line in output.splitlines()]
+		output, _ = dispatch_subprocess(['conan', 'search', query, '--remote', remote_name])
+		print(f'search result for {name}:\n{output}')
```
Comment (Jaguar0625) [inline jenkins/catapult/catapultDependencyUpdater.py:64]: did you leave print intentionally ?

File: jenkins/catapult/requirements.txt [hunk 8]
```diff
@@ -1,2 +1 @@
 PyYAML==6.0.1
```
Comment (Jaguar0625) [inline jenkins/catapult/requirements.txt:1]: don't you need conan package as dependency ?

**Review by Jaguar0625 — COMMENTED** [review]
👍

File: client/catapult/docs/BUILD-conan.md (hunk 3 above)
Comment (Wayonb) [reply client/catapult/docs/BUILD-conan.md:101]: Yeah 😅, but I left it since we always support the last two versions of the compiler.
We don't need both in the docs if you want to remove it.

File: client/catapult/CMakeLists.txt (hunk 6 above)
Comment (Wayonb) [reply client/catapult/CMakeLists.txt:121]: ``CatapultConan.generate`` copies all the dependencies libs from their Conan package to the catapult build folder.
When ``cmake install`` or ``ninja install`` is run, the above code will copy the deps folder from the catapult build folder to the install folder.

File: jenkins/catapult/catapultDependencyUpdater.py (hunk 7 above)
Comment (Wayonb) [reply jenkins/catapult/catapultDependencyUpdater.py:64]: no, can be removed

File: jenkins/catapult/requirements.txt (hunk 8 above)
Comment (Wayonb) [reply jenkins/catapult/requirements.txt:1]: Yeah, I miss this.

**Review by Jaguar0625 — CHANGES_REQUESTED** [review]
seems lint is failing 

seems the following files need to bump cmake_minimum:
* client/catapult/extensions/mongo/plugins/CMakeLists.txt
* client/catapult/extensions/mongo/tests/test/CMakeLists.txt

File: jenkins/catapult/requirements.txt [hunk 9]
```diff
@@ -1,2 +1,2 @@
...
-node-semver==0.6.1 # use the same version as conan to avoid conflicts
```
Comment (Jaguar0625) [inline jenkins/catapult/requirements.txt:2]: is this the right requirements or should client/catapult/scripts/requirements.txt also be updated ?

**Review by Jaguar0625 — COMMENTED** [review]
seems build is failing with error:

```
[2024-03-15T15:18:41.637Z] usage: conan profile [-h] {list,show,new,update,get,remove} ...
[2024-03-15T15:18:41.637Z] conan profile: error: unrecognized arguments: --profile
```

wrong version of conan?

File: jenkins/catapult/requirements.txt (hunk 9 above)
Comment (Wayonb) [reply jenkins/catapult/requirements.txt:2]: yea, it's needed here for the dependency updater scripts.
but probably should also be in ``client/catapult/scripts/requirements.txt``

**Comment by Wayonb** [comment]
Still need to rebuild the images

File: jenkins/catapult/runDockerBuildInnerBuild.py [hunk 10]
```diff
@@ -43,30 +42,35 @@ def _prepare_environment_variables(self):
...
-			self.dispatch_subprocess(['conan', 'profile', 'update', f'settings.compiler.{key}={value}', 'default'])
-
-	def run_conan_install(self, source_path):
+	def run_conan_install(self, source_path, settings, build_path, build_type):
 		# assuming working directory == build directory
-		self.dispatch_subprocess(['conan', 'profile', 'show', 'default'])
-		self.dispatch_subprocess(['conan', 'install', source_path, '--build', 'missing'])
+		setting_override = []
+		for key, value in settings.items():
+			setting_override = ['-s', f'compiler.{key}={value}']
```
Comment (Jaguar0625) [inline jenkins/catapult/runDockerBuildInnerBuild.py:57]: += ?

**Review by Jaguar0625 — APPROVED** [review]
👍

File: client/catapult/CMakeLists.txt [hunk 11]
```diff
@@ -48,6 +48,9 @@ if(ENABLE_TESTS)
 	message("GTest   found: ${GTest_FOUND}")
 	message("GTest     inc: ${GTest_INCLUDE_DIRS}")
 	message("GTest     ver: ${GTest_VERSION}")
+	if (NOT GTest_LIBRARIES AND GTEST_LIBRARIES)
```
Comment (Jaguar0625) [inline client/catapult/CMakeLists.txt:51]: add comment why you need this

**Review by Jaguar0625 — APPROVED** [review]
👍

**Review by Jaguar0625 — APPROVED** [review]
👍
